caraml-store-serving/src/main/java/dev/caraml/serving/store/bigtable/HBaseOnlineRetriever.java

```diff
@@ -0,0 +1,209 @@
+package dev.caraml.serving.store.bigtable;
+
+import com.google.protobuf.ByteString;
+import com.google.protobuf.Timestamp;
+import dev.caraml.serving.store.AvroFeature;
+import dev.caraml.serving.store.Feature;
+import dev.caraml.store.protobuf.serving.ServingServiceProto;
+import java.io.IOException;
+import java.nio.ByteBuffer;
+import java.util.*;
+import java.util.stream.Collectors;
+import org.apache.avro.AvroRuntimeException;
+import org.apache.avro.generic.GenericDatumReader;
+import org.apache.avro.generic.GenericRecord;
+import org.apache.avro.io.BinaryDecoder;
+import org.apache.avro.io.DecoderFactory;
+import org.apache.hadoop.hbase.Cell;
+import org.apache.hadoop.hbase.TableName;
+import org.apache.hadoop.hbase.client.*;
+
+public class HBaseOnlineRetriever implements SSTableOnlineRetriever<ByteString, Result> {
+  private final Connection client;
+  private final HBaseSchemaRegistry schemaRegistry;
+
+  public HBaseOnlineRetriever(Connection client) {
+    this.client = client;
+    this.schemaRegistry = new HBaseSchemaRegistry(client);
+  }
+
+  /**
+   * Generate Hbase key in the form of entity values joined by #.
+   *
+   * @param entityRow Single EntityRow representation in feature retrieval call
+   * @param entityNames List of entities related to feature references in retrieval call
+   * @return
+   */
+  @Override
+  public ByteString convertEntityValueToKey(
+      ServingServiceProto.GetOnlineFeaturesRequest.EntityRow entityRow, List<String> entityNames) {
+    return ByteString.copyFrom(
+        entityNames.stream()
+            .sorted()
+            .map(entity -> entityRow.getFieldsMap().get(entity))
+            .map(this::valueToString)
+            .collect(Collectors.joining("#"))
+            .getBytes());
+  }
+
+  /**
+   * Converts rowCell feature into @NativeFeature type, HBase specific implementation
+   *
+   * @param tableName Name of SSTable
+   * @param rowKeys List of keys of rows to retrieve
+   * @param rows Map of rowKey to Row related to it
+   * @param featureReferences List of feature references
+   * @return List of List of Features associated with respective rowKey
+   */
+  @Override
+  public List<List<Feature>> convertRowToFeature(
+      String tableName,
+      List<ByteString> rowKeys,
+      Map<ByteString, Result> rows,
+      List<ServingServiceProto.FeatureReference> featureReferences) {
+    BinaryDecoder reusedDecoder = DecoderFactory.get().binaryDecoder(new byte[0], null);
+
+    return rowKeys.stream()
+        .map(
+            rowKey -> {
+              if (!rows.containsKey(rowKey)) {
+                return Collections.<Feature>emptyList();
+              } else {
+                Result row = rows.get(rowKey);
+                return featureReferences.stream()
```

Comment:
the chain is very long, is it possible to break the chain?

Reply:
done at https://github.com/caraml-dev/caraml-store/pull/127/commits/4355a1a841a7534140af90882cfeeb9bb2b6c43f thank you 🙏